File: Readers/JavaScript/l.js
/**
 * Lightweight reader class for .locbook files.
 * Provides a simple API for reading localization data from JSON-based .locbook files.
 * 
 * Namespace equivalent: AHAKuo.Lingramia.API
 */

const fs = require('fs');
const path = require('path');

class L {
    /**
     * Creates a new L instance.
     */
    constructor() {
        this._resourcePath = null;
        this._currentLanguage = 'en';
        this._cache = {};
        this._keyIndex = {};
    }

    /**
     * Sets the resource path containing .locbook files.
     * Loads all .locbook files from the specified directory into cache.
     * 
     * @param {string} pathToResources - Path to the folder containing .locbook files
     * @throws {Error} If path is null or empty
     * @throws {Error} If directory does not exist
     */
    setResourcePath(pathToResources) {
        if (!pathToResources || !pathToResources.trim()) {
            throw new Error('Resource path cannot be null or empty.');
        }

        if (!fs.existsSync(pathToResources) || !fs.statSync(pathToResources).isDirectory()) {
            throw new Error(`Directory not found: ${pathToResources}`);
        }

        this._resourcePath = pathToResources;
        this._cache = {};
        this._keyIndex = {};

        // Load all .locbook files from the directory
        const files = fs.readdirSync(pathToResources);
        for (const filename of files) {
            if (filename.endsWith('.locbook')) {
                const filePath = path.join(pathToResources, filename);
                try {
                    this._loadLocbookFile(filePath);
                } catch (ex) {
                    // Log warning but don't crash - continue loading other files
                    console.warn(`Warning: Failed to load ${filePath}: ${ex.message}`);
                }
            }
        }
    }

    /**
     * Gets the current active language code.
     * 
     * @returns {string} The current language code (e.g., "en", "jp", "ar")
     */
    getLanguage() {
        return this._currentLanguage;
    }

    /**
     * Sets the active language for translations.
     * 
     * @param {string} code - Language code (e.g., "en", "jp", "ar")
     * @throws {Error} If code is null or empty
     */
    setLanguage(code) {
        if (!code || !code.trim()) {
            throw new Error('Language code cannot be null or empty.');
        }

        this._currentLanguage = code;
    }

    /**
     * Looks up a translation value by key.
     * 
     * @param {string} key - The key to look up
     * @param {boolean} hybridKey - If true, tries key, then originalValue, then aliases as fallback
     * @returns {string|null} The translated value for the current language, or null if not found
     */
    key(key, hybridKey = false) {
        if (!key || !key.trim()) {
            return null;
        }

        // Standard mode: lookup by key only
        if (!hybridKey) {
            return this._lookupByKey(key);
        }

        // Hybrid mode: try key, then originalValue, then aliases
        let result = this._lookupByKey(key);
        if (result !== null) {
            return result;
        }

        // Try originalValue
        const byOriginalValue = this._lookupByOriginalValue(key);
        if (byOriginalValue !== null) {
            return byOriginalValue;
        }

        // Try aliases
        const byAlias = this._lookupByAlias(key);
        if (byAlias !== null) {
            return byAlias;
        }

        // Fallback: return null
        return null;
    }

    /**
     * Internal method to lookup by key.
     * @private
     */
    _lookupByKey(key) {
        for (const fileName in this._keyIndex) {
            const pageFiles = this._keyIndex[fileName];
            if (pageFiles.hasOwnProperty(key)) {
                const pageFile = pageFiles[key];
                return this._getTranslationForLanguage(pageFile, this._currentLanguage);
            }
        }
        return null;
    }

    /**
     * Internal method to lookup by originalValue.
     * @private
     */
    _lookupByOriginalValue(originalValue) {
        for (const fileName in this._keyIndex) {
            const pageFiles = this._keyIndex[fileName];
            for (const key in pageFiles) {
                const pageFile = pageFiles[key];
                if (pageFile.originalValue && 
                    pageFile.originalValue.toLowerCase() === originalValue.toLowerCase()) {
                    return this._getTranslationForLanguage(pageFile, this._currentLanguage);
                }
            }
        }
        return null;
    }

    /**
     * Internal method to lookup by alias.
     * @private
     */
    _lookupByAlias(alias) {
        for (const fileName in this._keyIndex) {
            const pageFiles = this._keyIndex[fileName];
            for (const key in pageFiles) {
                const pageFile = pageFiles[key];
                const aliases = pageFile.aliases || [];
                if (aliases.some(a => a && a.toLowerCase() === alias.toLowerCase())) {
                    return this._getTranslationForLanguage(pageFile, this._currentLanguage);
                }
            }
        }
        return null;
    }

    /**
     * Internal method to get translation for a specific language.
     * @private
     */
    _getTranslationForLanguage(pageFile, language) {
        const variants = pageFile.variants || [];

        if (variants.length === 0) {
            // Fallback to original value
            return pageFile.originalValue || null;
        }

        // Find variant matching the language
        for (const variant of variants) {
            if (variant.language && variant.language.toLowerCase() === language.toLowerCase()) {
                const value = variant._value;
                if (value) {
                    return value;
                }
            }
        }

        // Fallback to original value if translation not found
        return pageFile.originalValue || null;
    }

    /**
     * Internal method to load a .locbook file into cache.
     * @private
     */
    _loadLocbookFile(filePath) {
        const json = fs.readFileSync(filePath, 'utf8');
        const locbook = JSON.parse(json);

        if (!locbook || !locbook.pages) {
            return;
        }

        const fileName = path.basename(filePath, '.locbook');
        this._cache[fileName] = locbook;

        // Build key index for fast lookup
        const pages = locbook.pages || [];
        for (const page of pages) {
            const pageFiles = page.pageFiles || [];
            if (pageFiles.length === 0) {
                continue;
            }

            if (!this._keyIndex[fileName]) {
                this._keyIndex[fileName] = {};
            }

            for (const pageFile of pageFiles) {
                const key = (pageFile.key || '').trim();
                if (key) {
                    this._keyIndex[fileName][key] = pageFile;
                }
            }
        }
    }
}

// Export for Node.js
if (typeof module !== 'undefined' && module.exports) {
    module.exports = L;
}

// Export for ES6 modules
if (typeof window !== 'undefined') {
    window.L = L;
}

File: Readers/Python/l.py
"""
Lightweight reader class for .locbook files.
Provides a simple API for reading localization data from JSON-based .locbook files.
"""

import json
import os
from typing import Optional, Dict, List, Any


class L:
    """
    Reader class for .locbook files.
    Namespace equivalent: AHAKuo.Lingramia.API
    """
    
    def __init__(self):
        self._resource_path: Optional[str] = None
        self._current_language: str = "en"
        self._cache: Dict[str, Dict[str, Any]] = {}
        self._key_index: Dict[str, Dict[str, Dict[str, Any]]] = {}
    
    def set_resource_path(self, path_to_resources: str) -> None:
        """
        Sets the resource path containing .locbook files.
        Loads all .locbook files from the specified directory into cache.
        
        Args:
            path_to_resources: Path to the folder containing .locbook files
            
        Raises:
            ValueError: If path is None or empty
            FileNotFoundError: If directory does not exist
        """
        if not path_to_resources or not path_to_resources.strip():
            raise ValueError("Resource path cannot be None or empty.")
        
        if not os.path.isdir(path_to_resources):
            raise FileNotFoundError(f"Directory not found: {path_to_resources}")
        
        self._resource_path = path_to_resources
        self._cache.clear()
        self._key_index.clear()
        
        # Load all .locbook files from the directory
        for filename in os.listdir(path_to_resources):
            if filename.endswith('.locbook'):
                file_path = os.path.join(path_to_resources, filename)
                try:
                    self._load_locbook_file(file_path)
                except Exception as ex:
                    # Log warning but don't crash - continue loading other files
                    print(f"Warning: Failed to load {file_path}: {ex}")
    
    def get_language(self) -> str:
        """
        Gets the current active language code.
        
        Returns:
            The current language code (e.g., "en", "jp", "ar")
        """
        return self._current_language
    
    def set_language(self, code: str) -> None:
        """
        Sets the active language for translations.
        
        Args:
            code: Language code (e.g., "en", "jp", "ar")
            
        Raises:
            ValueError: If code is None or empty
        """
        if not code or not code.strip():
            raise ValueError("Language code cannot be None or empty.")
        
        self._current_language = code
    
    def key(self, key: str, hybrid_key: bool = False) -> Optional[str]:
        """
        Looks up a translation value by key.
        
        Args:
            key: The key to look up
            hybrid_key: If True, tries key, then originalValue, then aliases as fallback
            
        Returns:
            The translated value for the current language, or None if not found
        """
        if not key or not key.strip():
            return None
        
        # Standard mode: lookup by key only
        if not hybrid_key:
            return self._lookup_by_key(key)
        
        # Hybrid mode: try key, then originalValue, then aliases
        result = self._lookup_by_key(key)
        if result is not None:
            return result
        
        # Try originalValue
        by_original_value = self._lookup_by_original_value(key)
        if by_original_value is not None:
            return by_original_value
        
        # Try aliases
        by_alias = self._lookup_by_alias(key)
        if by_alias is not None:
            return by_alias
        
        # Fallback: return None
        return None
    
    def _lookup_by_key(self, key: str) -> Optional[str]:
        """Internal method to lookup by key."""
        for file_name, page_files in self._key_index.items():
            if key in page_files:
                page_file = page_files[key]
                return self._get_translation_for_language(page_file, self._current_language)
        return None
    
    def _lookup_by_original_value(self, original_value: str) -> Optional[str]:
        """Internal method to lookup by originalValue."""
        for file_name, page_files in self._key_index.items():
            for page_file in page_files.values():
                if page_file.get('originalValue', '').lower() == original_value.lower():
                    return self._get_translation_for_language(page_file, self._current_language)
        return None
    
    def _lookup_by_alias(self, alias: str) -> Optional[str]:
        """Internal method to lookup by alias."""
        for file_name, page_files in self._key_index.items():
            for page_file in page_files.values():
                aliases = page_file.get('aliases', [])
                if aliases and any(a.lower() == alias.lower() for a in aliases if a):
                    return self._get_translation_for_language(page_file, self._current_language)
        return None
    
    def _get_translation_for_language(self, page_file: Dict[str, Any], language: str) -> Optional[str]:
        """Internal method to get translation for a specific language."""
        variants = page_file.get('variants', [])
        
        if not variants:
            # Fallback to original value
            return page_file.get('originalValue')
        
        # Find variant matching the language
        for variant in variants:
            if variant.get('language', '').lower() == language.lower():
                value = variant.get('_value')
                if value:
                    return value
        
        # Fallback to original value if translation not found
        return page_file.get('originalValue')
    
    def _load_locbook_file(self, file_path: str) -> None:
        """Internal method to load a .locbook file into cache."""
        with open(file_path, 'r', encoding='utf-8') as f:
            locbook = json.load(f)
        
        if not locbook or 'pages' not in locbook:
            return
        
        file_name = os.path.splitext(os.path.basename(file_path))[0]
        self._cache[file_name] = locbook
        
        # Build key index for fast lookup
        pages = locbook.get('pages', [])
        for page in pages:
            page_files = page.get('pageFiles', [])
            if not page_files:
                continue
            
            if file_name not in self._key_index:
                self._key_index[file_name] = {}
            
            for page_file in page_files:
                key = page_file.get('key', '').strip()
                if key:
                    self._key_index[file_name][key] = page_file

